pkgs/build-support/rust/replace-workspace-values.py

```diff
@@ -0,0 +1,84 @@
+# This script implements the workspace inheritance mechanism described
+# here: https://doc.rust-lang.org/cargo/reference/workspaces.html#the-package-table
+#
+# Please run `mypy --strict`, `black`, and `isort --profile black` on this after editing, thanks!
+
+import copy
+import sys
+from typing import Any
+
+import tomli
+import tomli_w
+
+
+def load_file(path: str) -> dict[str, Any]:
+    with open(path, "rb") as f:
+        return tomli.load(f)
+
+
+def replace_key(
+    workspace_manifest: dict[str, Any], table: dict[str, Any], section: str, key: str
+) -> None:
+    if "workspace" in table[key] and table[key]["workspace"] is True:
+        print("replacing " + key)
+
+        replaced = table[key]
+        del replaced["workspace"]
+
+        workspace_copy = workspace_manifest[section][key]
+
+        if section == "dependencies":
+            crate_features = replaced.get("features")
+
+            if type(workspace_copy) is str:
+                replaced["version"] = workspace_copy
+            else:
+                replaced.update(workspace_copy)
+
+                merged_features = (crate_features or []) + (
+                    workspace_copy.get("features") or []
+                )
+
+                if len(merged_features) > 0:
+                    # Dictionaries are guaranteed to be ordered (https://stackoverflow.com/a/7961425)
+                    replaced["features"] = list(dict.fromkeys(merged_features))
+        elif section == "package":
+            table[key] = replaced = workspace_copy
+
+
+def replace_dependencies(
+    workspace_manifest: dict[str, Any], root: dict[str, Any]
+) -> None:
+    for key in ["dependencies", "dev-dependencies", "build-dependencies"]:
+        if key in root:
+            for k in root[key].keys():
+                replace_key(workspace_manifest, root[key], "dependencies", k)
+
+
+def main() -> None:
+    crate_manifest = load_file(sys.argv[1])
+    workspace_manifest = load_file(sys.argv[2])["workspace"]
+
+    if "workspace" in crate_manifest:
+        return
+
+    orig_crate_manifest = copy.deepcopy(crate_manifest)
```

Comment:
is this really worth it?

Reply:
As an alternative, we should probably have a global that is set to true [here](https://github.com/NixOS/nixpkgs/blob/0cc4830f14578ece6a3e0c1e543785570e26f2e4/pkgs/build-support/rust/replace-workspace-values.py#L23), unconditionally `deepcopy`ing the whole table doesn't look like optimization to me, if that's what you were going for

Reply:
Fair, not sure what I was thinking there 😕

Reply:
(I ended up not using a global, but a bool returned from each function and ORed.)